src/ring_theory/class_group.lean

```diff
@@ -99,6 +99,55 @@ noncomputable def class_group.mk : (fractional_ideal R⁰ K)ˣ →* class_group
 (quotient_group.mk' (to_principal_ideal R (fraction_ring R)).range).comp
   (units.map (fractional_ideal.canonical_equiv R⁰ K (fraction_ring R)))
 
+lemma class_group.mk_eq_mk {I J : (fractional_ideal R⁰ $ fraction_ring R)ˣ} :
+  class_group.mk I = class_group.mk J
+    ↔ ∃ x : (fraction_ring R)ˣ, I * to_principal_ideal R (fraction_ring R) x = J :=
+begin
+  erw [quotient_group.mk'_eq_mk', canonical_equiv_self, units.map_id],
+  exact ⟨λ ⟨_, ⟨x, hx⟩, h⟩, ⟨x, hx.symm ▸ h⟩, λ ⟨_, h⟩, ⟨_, ⟨_, rfl⟩, h⟩⟩
+end
+
+lemma class_group.mk_eq_mk_of_coe_ideal
+  {I J : (fractional_ideal R⁰ $ fraction_ring R)ˣ} {I' J' : ideal R}
+  (hI : (I : fractional_ideal R⁰ $ fraction_ring R) = I')
+  (hJ : (J : fractional_ideal R⁰ $ fraction_ring R) = J') :
+  class_group.mk I = class_group.mk J
+    ↔ ∃ x y : R, x ≠ 0 ∧ y ≠ 0 ∧ ideal.span {x} * I' = ideal.span {y} * J' :=
+begin
+  rw [class_group.mk_eq_mk],
+  split,
+  { rintro ⟨x, rfl⟩,
+    rw [units.coe_mul, hI, coe_to_principal_ideal, mul_comm,
+        span_singleton_mul_coe_ideal_eq_coe_ideal] at hJ,
+    exact ⟨_, _, sec_fst_ne_zero le_rfl x.ne_zero, sec_snd_ne_zero le_rfl ↑x, hJ⟩ },
+  { rintro ⟨_, _, hx, hy, h⟩,
+    simp only [mul_comm] at h,
+    have inj := no_zero_smul_divisors.algebra_map_injective R (fraction_ring R),
+    exact ⟨mk0 _ ((map_ne_zero_iff _ inj).mpr hx) / mk0 _ ((map_ne_zero_iff _ inj).mpr hy),
+      by simp_rw [div_eq_mul_inv, _root_.map_mul, ← mul_assoc, _root_.map_inv,
+                  _root_.mul_inv_eq_iff_eq_mul, ext_iff, units.coe_mul, coe_to_principal_ideal,
+                  coe_mk0, ← coe_ideal_span_singleton, hI, hJ, ← coe_ideal_mul, h]⟩ }
+end
+
+lemma class_group.mk_eq_one_of_coe_ideal {I : (fractional_ideal R⁰ $ fraction_ring R)ˣ}
+  {I' : ideal R} (hI : (I : fractional_ideal R⁰ $ fraction_ring R) = I') :
+  class_group.mk I = 1 ↔ ∃ x : R, x ≠ 0 ∧ I' = ideal.span {x} :=
+begin
+  have h1 :
+    (↑(1 : (fractional_ideal R⁰ $ fraction_ring R)ˣ) : fractional_ideal R⁰ $ fraction_ring R)
+      = (⊤ : ideal R) := rfl,
+  rw [← map_one class_group.mk, class_group.mk_eq_mk_of_coe_ideal hI h1],
+  split,
+  { rintro ⟨x, _, hx, hy, h⟩,
```

Comment:
How you could avoid writing out the type:
```suggestion
  rw [← map_one class_group.mk, class_group.mk_eq_mk_of_coe_ideal hI (_ : _ = ↑⊤)],
  swap, refl,
  split,
  { rintro ⟨x, y, hx, hy, h⟩,
```
BTW why do you like to use an underscore for `y` here? It produces an ugly variable name in the goal view, while `y` is just one character ...

Reply:
My philosophy has been to only declare variables only when they're used, but you're right